--- T1Lib-1.2/lib/t1lib/t1finfo.c.orig	Fri May 25 18:28:46 2001
+++ T1Lib-1.2/lib/t1lib/t1finfo.c	Wed Sep  5 00:06:39 2001
@@ -510,9 +510,9 @@
   char *c_name;
   char cc_name[128];
   static char succ[MAX_LIGS];
-  char succ_index;
+  int succ_index;
   static char lig[MAX_LIGS];
-  char lig_index;
+  int lig_index;
   
   Ligature *ligs;
   int i,j;
--- T1Lib-1.2/lib/t1lib/t1subset.c.orig	Fri May 25 08:45:39 2001
+++ T1Lib-1.2/lib/t1lib/t1subset.c	Tue Sep  4 23:26:27 2001
@@ -434,7 +434,7 @@
      Each charstring must be written once, even if the respective 
      character appears more than once in the encoding. So we set up
      table to remember which charstrings already have been written. */
-  if (( csdone=(char *)calloc( pFontBase->pFontArray[FontID].pType1Data->CharStringsP[0].key.len,
+  if (( csdone=(char *)calloc( pFontBase->pFontArray[FontID].pType1Data->CharStringsP[0].key.len + 1,
 			       1))==NULL) {
     T1_errno=T1ERR_ALLOC_MEM;
     free( filebuf);
