--- T1Lib-1.2/lib/t1lib/t1finfo.c.orig	Fri May 25 18:28:46 2001
+++ T1Lib-1.2/lib/t1lib/t1finfo.c	Wed Sep  5 00:06:39 2001
@@ -510,9 +510,9 @@
   char *c_name;
   char cc_name[128];
   static char succ[MAX_LIGS];
-  char succ_index;
+  int succ_index;
   static char lig[MAX_LIGS];
-  char lig_index;
+  int lig_index;
   
   Ligature *ligs;
   int i,j;
--- T1Lib-1.2/lib/t1lib/t1subset.c.orig	Fri May 25 08:45:39 2001
+++ T1Lib-1.2/lib/t1lib/t1subset.c	Tue Sep 15 17:10:55 2001
@@ -453,7 +453,7 @@
   memcpy(&(filebuf[i]), charstringP, charstringL);
   i+=charstringL;
   i+=sprintf( &(filebuf[i]), " %s\n", ndstring);
-  csdone[currstring_no]=1;
+  csdone[currstring_no-1]=1;
   /* Now, step through the specifier matrix and write only the
      necessary charstrings. */
   for ( j=0; j<256; j++) {
